src/Psy/ExecutionLoop/Loop.php

```diff
@@ -27,17 +30,55 @@ class Loop
      */
     public function __construct(Configuration $config)
     {
-        // don't need this
+        $this->cleaner = $config->getCodeCleaner();
     }
 
     /**
-     * Run the execution loop.
-     *
-     * @param Shell $shell
+     * @return callable
      */
-    public function run(Shell $shell)
+    protected function getExecutionClosure()
     {
-        $loop = function ($__psysh__) {
+        return function (Shell &$__psysh__, $code = null) {
```

Comment:
Again, we're leaking a variable into local scope. This one's even more problematic, because it's *way* more likely that someone would have a variable called $code, which would be extracted on line 44 below and executed :-/

It's really looking like having a second code path just for `execute()` would be far cleaner.